File: src/main/java/tech/sirwellington/alchemy/http/AlchemyRequestMapper.java
/*
 * Copyright 2015 SirWellington Tech.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.http;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.charset.StandardCharsets;

import org.apache.http.HttpEntity;
import org.apache.http.client.methods.HttpDelete;
import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpPut;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.entity.StringEntity;
import tech.sirwellington.alchemy.annotations.access.Internal;
import tech.sirwellington.alchemy.annotations.arguments.Required;
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern;
import tech.sirwellington.alchemy.http.exceptions.AlchemyHttpException;

import static java.nio.charset.StandardCharsets.UTF_8;
import static tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern.Role.CLIENT;
import static tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern.Role.INTERFACE;
import static tech.sirwellington.alchemy.arguments.Arguments.checkThat;
import static tech.sirwellington.alchemy.http.HttpAssertions.notNullAndHasURL;

/**
 * {@code AlchemyRequestMappers} convert {@linkplain HttpRequest Alchemy Requests} into a
 * corresponding {@linkplain org.apache.http.HttpRequest Apache Request}.
 *
 * @author SirWellington
 */
@StrategyPattern(role = INTERFACE)
@Internal
interface AlchemyRequestMapper
{

    HttpUriRequest convertToApacheRequest(@Required HttpRequest alchemyRequest) throws AlchemyHttpException;

    @Internal
    static URL expandUrlFromRequest(@Required HttpRequest request) throws URISyntaxException, MalformedURLException
    {
        checkThat(request).is(notNullAndHasURL());

        if (!request.hasQueryParams())
        {
            return request.getUrl();
        }
        else
        {
            URI uri = request.getUrl().toURI();
            URIBuilder uriBuilder = new URIBuilder(uri);

            request.getQueryParams()
                    .forEach(uriBuilder::addParameter);

            return uriBuilder.build().toURL();
        }
    }

    @StrategyPattern(role = CLIENT)
    static final AlchemyRequestMapper GET = request ->
    {
        checkThat(request).is(notNullAndHasURL());

        try
        {
            URL url = expandUrlFromRequest(request);
            HttpGet get = new HttpGet(url.toURI());
            return get;
        }
        catch (Exception ex)
        {
            throw new AlchemyHttpException(request, "Could not convert to Apache GET Request", ex);
        }
    };

    static final AlchemyRequestMapper POST = request ->
    {
        checkThat(request).is(notNullAndHasURL());

        try
        {
            URL url = expandUrlFromRequest(request);
            HttpPost post = new HttpPost(url.toURI());

            if (request.hasBody())
            {
                HttpEntity entity = new StringEntity(request.getBody().toString(), UTF_8);
                post.setEntity(entity);
            }

            return post;
        }
        catch (Exception ex)
        {
            throw new AlchemyHttpException(request, "Could not convert to Apache POST Request", ex);
        }
    };

    static final AlchemyRequestMapper PUT = request ->
    {
        checkThat(request).is(notNullAndHasURL());

        try
        {
            URL url = expandUrlFromRequest(request);
            HttpPut put = new HttpPut(url.toURI());

            if (request.hasBody())
            {
                HttpEntity entity = new StringEntity(request.getBody().toString(), UTF_8);
                put.setEntity(entity);
            }

            return put;
        }
        catch (Exception ex)
        {
            throw new AlchemyHttpException(request, "Could not convert to Apache PUT Request", ex);
        }
    };

    static final AlchemyRequestMapper DELETE = request ->
    {
        checkThat(request).is(notNullAndHasURL());

        try
        {
            URL url = expandUrlFromRequest(request);
            if (request.hasBody())
            {
                //Custom Delete that allows a Body in the request
                HttpDeleteWithBody delete = new HttpDeleteWithBody(url.toURI());
                HttpEntity entity = new StringEntity(request.getBody().toString(), UTF_8);
                delete.setEntity(entity);
                return delete;
            }
            else
            {
                //The stock Apache Method
                return new HttpDelete(url.toURI());
            }
        }
        catch (Exception ex)
        {
            throw new AlchemyHttpException(request, "Could not convert to Apache DELETE Request", ex);
        }
    };

    /**
     * The stock {@linkplain HttpDelete Apache Delete Request} does not allow a Body. This one does.
     */
    @Internal
    class HttpDeleteWithBody extends HttpEntityEnclosingRequestBase
    {

        public static final String METHOD_NAME = new HttpDelete().getMethod();

        @Override
        public String getMethod()
        {
            return METHOD_NAME;
        }

        public HttpDeleteWithBody(final String uri)
        {
            super();
            setURI(URI.create(uri));
        }

        public HttpDeleteWithBody(final URI uri)
        {
            super();
            setURI(uri);
        }

        public HttpDeleteWithBody()
        {
            super();
        }

    }

}

File: src/main/java/tech/sirwellington/alchemy/http/AlchemyRequestMapper.kt
/*
 * Copyright © 2018. Sir Wellington.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.http

import io.mikael.urlbuilder.UrlBuilder
import org.slf4j.LoggerFactory
import tech.sirwellington.alchemy.annotations.access.Internal
import tech.sirwellington.alchemy.annotations.arguments.Required
import tech.sirwellington.alchemy.annotations.designs.patterns.FactoryMethodPattern
import tech.sirwellington.alchemy.annotations.designs.patterns.FactoryMethodPattern.Role.FACTORY_METHOD
import tech.sirwellington.alchemy.annotations.designs.patterns.FactoryMethodPattern.Role.PRODUCT
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern.Role
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern.Role.INTERFACE
import tech.sirwellington.alchemy.arguments.Arguments.checkThat
import tech.sirwellington.alchemy.http.HttpAssertions.validRequest
import tech.sirwellington.alchemy.http.exceptions.AlchemyHttpException
import tech.sirwellington.alchemy.http.exceptions.OperationFailedException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URISyntaxException
import java.net.URL

/**
 *
 * @author SirWellington
 */
@StrategyPattern(role = INTERFACE)
@Internal
internal interface AlchemyRequestMapper
{

    fun map(request: HttpRequest) : HttpURLConnection

    companion object
    {
        @Throws(AlchemyHttpException::class, URISyntaxException::class, MalformedURLException::class)
        fun expandUrlFromRequest(@Required request: HttpRequest): URL
        {
            checkThat(request).isA(validRequest())

            val url = request.url ?: throw OperationFailedException("request is missing URL")

            return if (!request.hasQueryParams())
            {
                url
            }
            else
            {
                val uriBuilder = UrlBuilder.fromUrl(url)

                request.queryParams?.
                        forEach { param, value -> uriBuilder.addParameter(param, value) }

                uriBuilder.toUrl()
            }
        }

        @FactoryMethodPattern(role = FACTORY_METHOD)
        fun create(): AlchemyRequestMapper
        {
            return AlchemyRequestMapperImpl
        }
    }
}

@FactoryMethodPattern(role = PRODUCT)
@StrategyPattern(role = Role.CONCRETE_BEHAVIOR)
private object AlchemyRequestMapperImpl: AlchemyRequestMapper
{

    private val LOG = LoggerFactory.getLogger(this::class.java)

    override fun map(request: HttpRequest): HttpURLConnection
    {
        val url = AlchemyRequestMapper.expandUrlFromRequest(request)
        val connection = url.openConnection()

        val http = connection as? HttpURLConnection ?: throw OperationFailedException("URL is not an HTTP URL: [$url]")

        http.doInput = true
        http.requestMethod = request.method.asString

        request.requestHeaders?.forEach { key, value -> http.setRequestProperty(key, value) }

        if (request.hasBody())
        {
            http.doOutput = true
            http.setBody(request)
        }

        return http
    }

    private fun HttpURLConnection.setBody(request: HttpRequest)
    {
        val jsonString = request.body?.toString() ?: return

        try
        {
            this.outputStream.use { it ->
                it.bufferedWriter(Charsets.UTF_8).write(jsonString)
            }
        }
        catch (ex: Exception)
        {
            LOG.error("Failed to set json request body [{}]", jsonString, ex)
            throw OperationFailedException(request, "Failed to set json request body [$jsonString]", ex)
        }
    }
}